RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs

```diff
@@ -19,21 +20,26 @@ public PassParameterByReferenceQuickFix(ParserRuleContext context, QualifiedSele
 
         public override void Fix()
         {
-            string pattern = "^\\s*" + Tokens.ByVal + "(\\s+)";
-            string rgxReplacement = Tokens.ByRef + "$1";
-            Regex rgx = new Regex(pattern);
+            var byValParameter = Context.GetText();
 
-            var parameter = Context.GetText();
-            var newParameter = rgx.Replace(parameter, rgxReplacement);
+            var byRefParameter = BuildByRefParameter(byValParameter);
 
-            var selection = Selection.Selection;
+            ReplaceByValParameterInModule(byValParameter, byRefParameter);
+        }
 
+        private string BuildByRefParameter(string originalParameter)
+        {
+            var everythingAfterTheByValToken = originalParameter.Substring(Tokens.ByVal.Length);
+            return Tokens.ByRef + everythingAfterTheByValToken;
+        }
+        private void ReplaceByValParameterInModule( string byValParameter, string byRefParameter)
+        {
+            var selection = Selection.Selection;
             var module = Selection.QualifiedName.Component.CodeModule;
-            {
-                var lines = module.GetLines(selection.StartLine, selection.LineCount);
-                var result = lines.Replace(parameter, newParameter);
-                module.ReplaceLine(selection.StartLine, result);
-            }
+
+            var lines = module.GetLines(selection.StartLine, selection.LineCount);
+            var result = lines.Replace(byValParameter, byRefParameter);
```

Comment:
This. We need a code-rewriting API... like, *20 releases ago!*

We've indeed used regex in lots of places before we started *using* the Antlr parse trees - at some point between 1.4.3 and 2.0 the entire grammar was rewritten, which fixed a *ton* of issues.. but regex lingered, scattered throughout the earlier features.

Basically you need to rewrite an entire logical line of code (which could very well be more than just one *physical* line of code), and the only 100% safe way to do that is, for now, to work with the tokens.

Picture the user writing such weaponized test code:

    Sub DoSomething( _
        ByVal foo As Long, _
        ByVal _
            bar _
        ByRef barbecue
                        )
        bar = barbecue * _
                   bar + foo / barbecue
    End Sub

With the parse tree tokens you can surgically go and edit the *logical* line of code:

```csharp
    var context = Context as ParameterDeclaration;
    Debug.Asser(context != null && context.BYREF() == null);

    if (context.BYVAL() != null)
    {
        // parameter is passed by value
        // BYVAL token start & end positions are known,
        // so we can extract it from the string with then.
    }
    
    // need to insert the BYREF token just before the identifier token
    // use the identifier token's position to determine where to insert it.
````

It's a very "both hands dirty" approach, playing directly with token positions and the code module contents - we definitely need a nice layer of abstraction there.. but until we come up with that API, we need to work at that low abstraction level with the parse trees.